File: src/main/java/org/datastructures/thermometernetwork/ThermostatManager.java
package org.datastructures.thermometernetwork;

import java.util.*;
import java.io.*;
import java.text.SimpleDateFormat;

public class ThermostatManager {

    HashSet<Thermostat> thermostats = new HashSet<Thermostat>();

    public ThermostatManager() {
        // Intentionally empty
    }

    public boolean addThermostat(Thermostat thermostat) {
        return thermostats.add(thermostat);
    }

    public boolean addSensor(Thermostat thermostat, Sensor sensor) {
        return thermostat.addSensor(sensor);
    }

    // Only method that should have to change for compatibility with other systems
    // Writes to the save file in our format; does NOT create objects or add to internal data structure
    public static void saveNewData(String inFileName, String saveFileName) throws IOException {
        File inFile = new File(inFileName);
        File saveFile = new File(saveFileName);
        String line;
        String[] tokens;
        boolean even = true;

        try (Scanner in = new Scanner(inFile);
             FileWriter fileWriter = new FileWriter(saveFile, true);
             PrintWriter out = new PrintWriter(fileWriter)) {

            // Write metadata for thermostats/sensors if the save file is empty. Format:
            // thermostat1Name,sensor1name,sensor2name...\n
            // thermostat1Name,sensor1name,sensor2name...
            if (saveFile.length() == 0) {
                out.println("East Side,Master Bedroom,Master Bathroom,Office," +
                        "Ty's Bedroom,Luke's Bedroom,Sabry's Bedroom,Living Room\n" +
                        "West Side,Brody's Bedroom,Kitchen,Den,Bar,Living Room,Outside");
            }

            while (in.hasNextLine()) {
                line = in.nextLine();
                tokens = line.split("[\t ]+"); //Splits on tabs, spaces or multiples of tabs/spaces

                if (even) { // Only prints timestamp once per two lines of input
                    out.print(epochToDateTime(tokens[1]));
                }

                for (int i = 4; i < tokens.length; i++) {

                    if (tokens[2].equals("0")) { // East Side
                        switch (i) { // Ignores inactive sensors on East thermostat
                            case 4, 5, 6, 7, 8, 9, 11:
                                out.print("," + tokens[i]);
                                break;
                        }
                    }

                    else if (tokens[2].equals("1")) { // West Side
                        switch (i) { // Ignores inactive sensors on West thermostat
                            case 4, 5, 6, 7, 8, 11:
                                out.print("," + tokens[i]);
                                break;
                        }
                    }

                }// for-loop

                if (!even) { //Prints a \n once for every two lines of input
                    out.println();
                    even = true;
                }
                else {
                    even = false;
                }

            }// while-loop
        }// try-with-resources (automatically closes scanner and printWriter)
    }// method

    // Converts epoch to MM/DD/YYYY-hh:mm:ss
    public static String epochToDateTime(String epochTimestampStr) {
        double epochTimestamp = Double.parseDouble(epochTimestampStr);
        long epochMillis = (long) (epochTimestamp * 1000);
        Date date = new Date(epochMillis);// Create a Date object
        SimpleDateFormat dateFormat = new SimpleDateFormat("MM/dd/yyyy-HH:mm:ss");// Define the desired format
        return dateFormat.format(date);// Format the date and time
    }

    public void readSave(String saveFileName) throws IOException {
        File inFile = new File(saveFileName);
        String line;
        String[] firstDataLine;
        double temperature;
        String[] tokens;
        boolean isMetaData = true;

        try (Scanner in = new Scanner(inFile)) {

            //Reads metadata and creates Thermostat and Sensor objects
            while (in.hasNextLine()) {
                line = in.nextLine();
                tokens = line.split(",");

                // Regex pattern for the date/time format
                String dateTimeRegex = "^\\d{2}/\\d{2}/\\d{4}-\\d{2}:\\d{2}:\\d{2}$";

                if (isMetaData && !tokens[0].matches(dateTimeRegex)) { //This line is metadata
                    Thermostat thermostat = new Thermostat(tokens[0]);
                    addThermostat(thermostat);
                    for (int i = 1; i < tokens.length; i++) {
                        Sensor sensor = new Sensor(tokens[i]);
                        addSensor(thermostat, sensor); // Adds the current sensor to the current thermostat
                    }
                }

                else {
                    isMetaData = false;
                    //process real data. going to be aggravating I think.
                }
            }
        }
    }

}

File: backend/thermostat_manager.py
from thermostat import Thermostat
from sensor import Sensor
from ordered_set import OrderedSet
from collections import OrderedDict
from datetime import date
import os
from datetime import datetime
import re

class ThermostatManager:

    def __init__(self):
        self.thermostats = OrderedSet()
    
    def addThermostat(self, thermostat):
        return self.thermostats.add(thermostat)
    
    def addSensor(self, thermostat, sensor):
        return thermostat.addSensor(sensor)
    
    def getThermostats(self):
        return self.thermostats
    
    def writeToSave(self, inFileName, saveFileName): 
        try:
            inFile = open(inFileName, "r")
            saveFile = open(saveFileName, "w")
            even = True

            fileSize = os.path.getsize(saveFileName)
            
            if fileSize == 0:
                saveFile.write("East Side,Master Bedroom,Master Bathroom,Office," +
                        "Ty's Bedroom,Luke's Bedroom,Sabry's Bedroom,Living Room\n" +
                        "West Side,Brody's Bedroom,Kitchen,Den,Bar,Living Room,Outside")
            
            line = inFile.readline()
            tokens = line.split("[\t ]+")

            if even:
                saveFile.write(self.epochToDateTime(tokens[1]))

            for i in range(4, tokens.__len__):
                if tokens[2] == "0":
                    if i  != 5 and i != 12 and i != 14 and i != 15:
                        saveFile.write("," + tokens[i])
                        break

                elif tokens[2] == "1":
                    if i  != 5 and i != 11 and i != 12 and i != 14 and i != 15:
                        saveFile.write("," + tokens[i])
                        break
            
            if even:
                even = False
            else:
                saveFile.write("\n")
                even = True
            
            while line:
                line = inFile.readline()
                tokens = line.split("[\t ]+")

                if even:
                    saveFile.write(self.epochToDateTime(tokens[1]))

                for i in range(4, tokens.__len__):
                    if tokens[2] == "0":
                        if i  != 5 and i != 12 and i != 14 and i != 15:
                            saveFile.write("," + tokens[i])
                            break

                    elif tokens[2] == "1":
                        if i  != 5 and i != 11 and i != 12 and i != 14 and i != 15:
                            saveFile.write("," + tokens[i])
                            break
            if even:
                even = False
            else:
                saveFile.write("\n")
                even = True

            saveFile.close()
            inFile.close()
        except FileNotFoundError:
            return False


    def epochToDateTime(self, epochTimestamp):
        dt_object = datetime.utcfromtimestamp(epochTimestamp)
        formatted_time = dt_object.strftime('%M/%d/%y-%H:%m:%s')
        return formatted_time
    
    def loadSave(self, saveFileName):
        isMetaData = True

        with open(saveFileName, 'r') as file:
            for line in file:
                tokens = line.split(",")

                dateTimeRegex = "^\\d{2}/\\d{2}/\\d{4}-\\d{2}:\\d{2}:\\d{2}$"

                if isMetaData and not re.match(dateTimeRegex, tokens[0]):
                    thermostat = Thermostat(tokens[0])
                    self.addThermostat(thermostat)

                    for token in tokens[1:]:
                        sensor = Sensor(token)
                        self.addSensor(thermostat, sensor)
                
                else:
                    temperatureIndex = 1
                    isMetaData = False
                    timestamp = tokens[0]
                    for thermostat in self.thermostats:
                        for sensors in thermostat.getSensors():
                            thermostat.addTemperature(sensor, timestamp, float(tokens[temperatureIndex]))
                            temperatureIndex += 1
                            
    def wipeSave(self, saveFileName):
        open(saveFileName, "w").close

    def __str__(self):
        toString = ""
        for thermostat in self.thermostats:
            toString += thermostat
        return toString